### PR TITLE
[BUGFIX] Media selection overlay for sulu version >= 1.3.0

### PR DESCRIPTION
**Pull request by turbo-ele**
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes none
| Related issues/PRs | none
| License | MIT

#### What's in this PR?

Media selection overlay has been moved in version 1.3 and therefore needed to be adapted in product's media tab.

From now on only sulu version >= 1.3 is supported by product-bundle.

Also the structure of product's javascript component was changed. All content tabs are now located in products/components/edit.

### CODE REVIEW

File: Resources/public/js/components/products/components/edit/media/main.js [hunk 1]
```diff
@@ -17,6 +17,17 @@ define(['config', 'suluproduct/util/product-delete-dialog'], function(Config, De
...
+
+    /**
+     * Returns a comparable version number by parsing a version string.
+     *
+     * @param {String} versionString
+     */
+    getVersionFromString = function(versionString) {
+        var versionParts = versionString.split('.');
+
+        return parseInt(versionParts.slice(0,3).join(''));
```
Comment (wachterjohannes) [inline Resources/public/js/components/products/components/edit/media/main.js]: space between , and 3

File: Resources/public/js/components/products/components/edit/media/main.js [hunk 2]
```diff
@@ -47,8 +59,7 @@ define(['config', 'suluproduct/util/product-delete-dialog'], function(Config, De
 
         bindCustomEvents: function() {
             this.sandbox.on('product.state.change', function(status) {
-                if (!this.options.data ||
-                    !this.options.data.attributes.status ||
+                if (!this.options.data || !this.options.data.attributes.status ||
```
Comment (wachterjohannes) [inline Resources/public/js/components/products/components/edit/media/main.js]: `||` into next line and if one will break down then all terms should go into a new line

File: Resources/public/js/components/products/components/edit/media/main.js [hunk 3]
```diff
@@ -279,18 +294,24 @@ define(['config', 'suluproduct/util/product-delete-dialog'], function(Config, De
...
         /**
-         * Starts the overlay-component responsible for selecting the documents
+         * Starts the overlay-component responsible for selecting the documents.
          */
         startSelectionOverlay: function() {
             var $container = this.sandbox.dom.createElement('<div/>');
             this.sandbox.dom.append(this.$el, $container);
 
+            var componentName = 'media-selection/overlay@sulumedia';
+            if (getVersionFromString(this.config['sulu_version']) < getVersionFromString('1.3.0')) {
```
Comment (wachterjohannes) [inline Resources/public/js/components/products/components/edit/media/main.js]: sulu_version is not safe here try to find another solution.

File: Resources/public/js/components/products/components/edit/media/main.js [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+define([
+        'suluproduct/util/product-delete-dialog',
+        'services/product-media-manager'
+    ],
```
Comment (wachterjohannes) [inline Resources/public/js/components/products/components/edit/media/main.js]: move function into this line then you can omit one tab for the whole component

File: Resources/public/js/components/products/main.js [hunk 5]
```diff
@@ -193,76 +193,68 @@ define([
...
-                if (mediaIndex > -1) {
-                    this.product.get('media').splice(mediaIndex, 1);
-                    mediaIds.splice(mediaIndex, 1);
-                }
-            }
+            var mediaIds = this.sandbox.util.arrayGetColumn(media, 'id');
+
+            this.processAjaxForMedia(mediaIds, productId);
+
+            //this.processAjaxForMedia(removedMediaIds, productId, 'DELETE');
```
Comment (wachterjohannes) [inline Resources/public/js/components/products/main.js]: ??? is this outcommented code necessary at this point?

File: Resources/public/js/components/products/main.js [hunk 6]
```diff
@@ -308,7 +302,8 @@ define([
             }
 
             this.sandbox.emit('sulu.router.navigate', 'pim/products/' + localization + '/edit:' + id + '/' + tabName);
-        },
+        }
```
Comment (wachterjohannes) [inline Resources/public/js/components/products/main.js]: ? strange format in the whole file

File: CHANGELOG.md [hunk 7]
```diff
@@ -2,6 +2,9 @@ CHANGELOG for Sulu Product Bundle
 =================================
 
 * dev-develop
+
```
Comment (wachterjohannes) [inline CHANGELOG.md]: remove this empty line

File: Controller/ProductMediaController.php [hunk 8]
```diff
@@ -26,74 +30,58 @@
...
-     * Returns the media manager.
+     * @param Request $request
      *
-     * @return MediaManagerInterface
+     * TODO: SchemaValidation
+     *
+     * @return Response
      */
-    protected function getMediaManager()
+    public function fieldsAction(Request $request)
```
Comment (wachterjohannes) [inline Controller/ProductMediaController.php]: `getFieldsAction`

File: Product/ProductMediaManager.php [hunk 9]
```diff
@@ -36,35 +39,125 @@ class ProductMediaManager implements ProductMediaManagerInterface
     protected $productEntityName;
 
     /**
+     * @var MediaRepositoryInterface
```
Comment (wachterjohannes) [inline Product/ProductMediaManager.php]: class comment missing

File: Product/ProductMediaManagerInterface.php [hunk 10]
```diff
@@ -11,14 +11,34 @@
 
 namespace Sulu\Bundle\ProductBundle\Product;
 
+use Sulu\Bundle\ProductBundle\Entity\ProductInterface;
 use Sulu\Component\Rest\ListBuilder\Doctrine\FieldDescriptor\DoctrineFieldDescriptor;
 
 interface ProductMediaManagerInterface
```
Comment (wachterjohannes) [inline Product/ProductMediaManagerInterface.php]: interface comment missing

File: Product/ProductMediaManager.php (hunk 9 above)
Comment (turbo-ele) [inline Product/ProductMediaManager.php]: done

**Comment by turbo-ele** [comment]
@wachterjohannes implemented reuested changes.

**Comment by ISC3141** [comment]
Hello, thanks for your nice product bundle.
Do you know if this bug fix will be released soon?
I would like to add some media to the products but it does not work with Sulu 1.4.5 and Product Bundle 0.16.5...

**Comment by turbo-ele** [comment]
@ISC3141 thanks for your feedback. we will have a look into the issue within the next few days.

**Pull request merged**